File: common/src/main/java/net/shiroha233/roadweaver/client/map/render/MapRenderers.java
package net.shiroha233.roadweaver.client.map.render;

import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.PlayerFaceRenderer;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;

import net.shiroha233.roadweaver.core.model.StructureConnection;

import java.util.List;
import java.util.function.IntUnaryOperator;

/**
 * 地图渲染器集合
 */
public final class MapRenderers {
    private MapRenderers() {}

    public interface SegmentInView { boolean test(int x1, int z1, int x2, int z2); }

    public static void renderGrid(GuiGraphics g,
                                  Font font,
                                  int mapX, int mapY, int mapW, int mapH,
                                  int innerPad,
                                  double viewMinX, double viewMaxX, double viewMinZ, double viewMaxZ,
                                  int colorGrid, int gridTargetPx, int colorText) {
        GridRenderer.render(g, font, mapX, mapY, mapW, mapH, innerPad,
                viewMinX, viewMaxX, viewMinZ, viewMaxZ, colorGrid, gridTargetPx, colorText);
    }

    public static void renderStructures(GuiGraphics g,
                                        List<BlockPos> points,
                                        IntUnaryOperator toScreenX,
                                        IntUnaryOperator toScreenY,
                                        java.util.function.BiPredicate<Integer, Integer> isInViewWorld,
                                        int size,
                                        int color,
                                        int left, int top, int right, int bottom) {
        for (BlockPos p : points) {
            int bx = p.getX();
            int bz = p.getZ();
            if (!isInViewWorld.test(bx, bz)) continue;
            int x = toScreenX.applyAsInt(bx);
            int y = toScreenY.applyAsInt(bz);
            drawFilledCircle(g, x, y, Math.max(1, size / 2), color, left, top, right, bottom);
        }
    }

    public static void renderConnections(GuiGraphics g,
                                         List<StructureConnection> connections,
                                         SegmentInView segmentInView,
                                         IntUnaryOperator toScreenX,
                                         IntUnaryOperator toScreenY,
                                         int thickness,
                                         int colorPlanned, int colorGenerating, int colorCompleted, int colorFailed,
                                         int left, int top, int right, int bottom) {
        for (StructureConnection c : connections) {
            int fx = c.from().getX();
            int fz = c.from().getZ();
            int tx = c.to().getX();
            int tz = c.to().getZ();
            if (!segmentInView.test(fx, fz, tx, tz)) continue;
            int x1 = toScreenX.applyAsInt(fx);
            int y1 = toScreenY.applyAsInt(fz);
            int x2 = toScreenX.applyAsInt(tx);
            int y2 = toScreenY.applyAsInt(tz);
            switch (c.status()) {
                case PLANNED -> {
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, colorPlanned, thickness, left, top, right, bottom);
                }
                case GENERATING -> {
                    RenderUtils.drawThickDashedLine(g, x1, y1, x2, y2, colorGenerating, thickness, 8, 6, left, top, right, bottom);
                }
                case COMPLETED -> {
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, colorCompleted, thickness, left, top, right, bottom);
                }
                case FAILED -> {
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, colorFailed, thickness, left, top, right, bottom);
                }
            }
        }
    }

    public static void renderRoadPolylines(GuiGraphics g,
                                           List<List<BlockPos>> polylines,
                                           SegmentInView segmentInView,
                                           IntUnaryOperator toScreenX,
                                           IntUnaryOperator toScreenY,
                                           int thickness,
                                           int color,
                                           int left, int top, int right, int bottom,
                                           int lodStepBlocks) {
        for (List<BlockPos> pl : polylines) {
            if (pl == null || pl.size() < 2) continue;
            BlockPos keep = pl.get(0);
            for (int i = 1; i < pl.size(); i++) {
                BlockPos curr = pl.get(i);
                int dx = curr.getX() - keep.getX();
                int dz = curr.getZ() - keep.getZ();
                int adx = Math.abs(dx);
                int adz = Math.abs(dz);
                int manhattan = adx + adz;
                if (manhattan < lodStepBlocks) continue;
                int x1w = keep.getX();
                int z1w = keep.getZ();
                int x2w = curr.getX();
                int z2w = curr.getZ();
                if (segmentInView.test(x1w, z1w, x2w, z2w)) {
                    int x1 = toScreenX.applyAsInt(x1w);
                    int y1 = toScreenY.applyAsInt(z1w);
                    int x2 = toScreenX.applyAsInt(x2w);
                    int y2 = toScreenY.applyAsInt(z2w);
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, color, thickness, left, top, right, bottom);
                }
                keep = curr;
            }
            BlockPos tail = pl.get(pl.size() - 1);
            if (tail != keep) {
                int x1w = keep.getX();
                int z1w = keep.getZ();
                int x2w = tail.getX();
                int z2w = tail.getZ();
                if (segmentInView.test(x1w, z1w, x2w, z2w)) {
                    int x1 = toScreenX.applyAsInt(x1w);
                    int y1 = toScreenY.applyAsInt(z1w);
                    int x2 = toScreenX.applyAsInt(x2w);
                    int y2 = toScreenY.applyAsInt(z2w);
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, color, thickness, left, top, right, bottom);
                }
            }
        }
    }

    public static void renderLegend(GuiGraphics g,
                                    Font font,
                                    int rightBound,
                                    int startY,
                                    int gap,
                                    int colorText,
                                    int colorStruct,
                                    int colorPlanned,
                                    int colorGenerating,
                                    int colorCompleted,
                                    int colorFailed,
                                    int structuresCount,
                                    int plannedCount,
                                    int generatingCount,
                                    int completedCount,
                                    int failedCount) {
        int y = startY;
        Component l1 = Component.translatable("gui.roadweaver.map.legend.structures").append(": ").append(Integer.toString(structuresCount));
        int w1 = font.width(l1);
        int x1 = rightBound - w1;
        int sr = x1 - gap;
        g.fill(sr - 5, y + 1, sr, y + 6, colorStruct);
        g.drawString(font, l1, x1, y, colorText, false);

        y += 16;
        Component l2 = Component.translatable("gui.roadweaver.map.legend.planned").append(": ").append(Integer.toString(plannedCount));
        int w2 = font.width(l2);
        int x2 = rightBound - w2;
        sr = x2 - gap;
        g.fill(sr - 28, y + 2, sr, y + 7, colorPlanned);
        g.drawString(font, l2, x2, y, colorText, false);

        y += 16;
        Component l3 = Component.translatable("gui.roadweaver.map.legend.generating").append(": ").append(Integer.toString(generatingCount));
        int w3 = font.width(l3);
        int x3 = rightBound - w3;
        sr = x3 - gap;
        int cy = y + 4;
        RenderUtils.drawThickDashedLine(g, sr - 28, cy, sr, cy, colorGenerating, 5, 8, 6, sr - 28, y + 1, sr, y + 8);
        g.drawString(font, l3, x3, y, colorText, false);

        y += 16;
        Component l4 = Component.translatable("gui.roadweaver.map.legend.completed").append(": ").append(Integer.toString(completedCount));
        int w4 = font.width(l4);
        int x4 = rightBound - w4;
        sr = x4 - gap;
        g.fill(sr - 28, y + 2, sr, y + 7, colorCompleted);
        g.drawString(font, l4, x4, y, colorText, false);

        y += 16;
        Component l5 = Component.translatable("gui.roadweaver.map.legend.failed").append(": ").append(Integer.toString(failedCount));
        int w5 = font.width(l5);
        int x5 = rightBound - w5;
        sr = x5 - gap;
        g.fill(sr - 28, y + 2, sr, y + 7, colorFailed);
        g.drawString(font, l5, x5, y, colorText, false);
    }

    private static void drawFilledCircle(GuiGraphics g,
                                         int centerX,
                                         int centerY,
                                         int radius,
                                         int color,
                                         int left,
                                         int top,
                                         int right,
                                         int bottom) {
        int r = Math.max(1, radius);
        int r2 = r * r;
        for (int dy = -r; dy <= r; dy++) {
            int y = centerY + dy;
            if (y < top || y > bottom) continue;
            int dxLimit = (int) Math.floor(Math.sqrt(r2 - dy * dy));
            int x0 = Math.max(left, centerX - dxLimit);
            int x1 = Math.min(right, centerX + dxLimit);
            if (x0 <= x1) {
                g.fill(x0, y, x1 + 1, y + 1, color);
            }
        }
    }

    public static void drawPlayerAvatar(GuiGraphics g,
                                        ResourceLocation skinTexture,
                                        int sx,
                                        int sy,
                                        float yawDeg,
                                        int size,
                                        int frameColor,
                                        int directionColor,
                                        int directionOutlineColor,
                                        int directionTipLen,
                                        int directionBaseOffset,
                                        int directionHalfWidth,
                                        int left,
                                        int top,
                                        int right,
                                        int bottom) {
        if (skinTexture == null) return;
        double rx = Math.toRadians(yawDeg);
        double dirX = -Math.sin(rx);
        double dirY = Math.cos(rx);
        double perpX = -dirY;
        double perpY = dirX;
        double tipX = sx + dirX * directionTipLen;
        double tipY = sy + dirY * directionTipLen;
        double baseX = sx + dirX * directionBaseOffset;
        double baseY = sy + dirY * directionBaseOffset;
        int ipx = (int) Math.round(tipX);
        int ipy = (int) Math.round(tipY);
        int ibx1 = (int) Math.round(baseX + perpX * directionHalfWidth);
        int iby1 = (int) Math.round(baseY + perpY * directionHalfWidth);
        int ibx2 = (int) Math.round(baseX - perpX * directionHalfWidth);
        int iby2 = (int) Math.round(baseY - perpY * directionHalfWidth);
        RenderUtils.fillTriangle(g, ipx - 1, ipy, ibx1 - 1, iby1, ibx2 - 1, iby2, directionOutlineColor, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx + 1, ipy, ibx1 + 1, iby1, ibx2 + 1, iby2, directionOutlineColor, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx, ipy - 1, ibx1, iby1 - 1, ibx2, iby2 - 1, directionOutlineColor, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx, ipy + 1, ibx1, iby1 + 1, ibx2, iby2 + 1, directionOutlineColor, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx, ipy, ibx1, iby1, ibx2, iby2, directionColor, left, top, right, bottom);

        int half = size / 2;
        int drawX = sx - half;
        int drawY = sy - half;
        if (drawX > right || drawX + size < left || drawY > bottom || drawY + size < top) {
            return;
        }
        g.fill(drawX - 2, drawY - 2, drawX + size + 2, drawY + size + 2, frameColor);
        g.fill(drawX - 1, drawY - 1, drawX + size + 1, drawY + size + 1, 0x66FFFFFF);
        PlayerFaceRenderer.draw(g, skinTexture, drawX, drawY, size);
    }

    public static void drawPlayerArrow(GuiGraphics g,
                                       int sx, int sy,
                                       float yawDeg,
                                       int tipLen,
                                       int baseLen,
                                       int baseHalfWidth,
                                       int color,
                                       int left, int top, int right, int bottom,
                                       double pxPerBlockX,
                                       double pxPerBlockZ) {
        double rx = Math.toRadians(yawDeg);
        double vx = -Math.sin(rx);
        double vz = Math.cos(rx);
        double dirX = vx * pxPerBlockX;
        double dirY = vz * pxPerBlockZ;
        double mag = Math.hypot(dirX, dirY);
        if (mag < 1e-6) { dirX = 1; dirY = 0; mag = 1; }
        dirX /= mag; dirY /= mag;
        double px = sx + dirX * tipLen;
        double py = sy + dirY * tipLen;
        double bx = sx - dirX * baseLen;
        double by = sy - dirY * baseLen;
        double perpX = -dirY;
        double perpY = dirX;
        double bx1 = bx + perpX * baseHalfWidth;
        double by1 = by + perpY * baseHalfWidth;
        double bx2 = bx - perpX * baseHalfWidth;
        double by2 = by - perpY * baseHalfWidth;
        int ipx = (int)Math.round(px);
        int ipy = (int)Math.round(py);
        int ibx1 = (int)Math.round(bx1);
        int iby1 = (int)Math.round(by1);
        int ibx2 = (int)Math.round(bx2);
        int iby2 = (int)Math.round(by2);
        int outline = 0xFFFFFFFF;
        RenderUtils.fillTriangle(g, ipx - 1, ipy,     ibx1 - 1, iby1,     ibx2 - 1, iby2,     outline, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx + 1, ipy,     ibx1 + 1, iby1,     ibx2 + 1, iby2,     outline, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx,     ipy - 1, ibx1,     iby1 - 1, ibx2,     iby2 - 1, outline, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx,     ipy + 1, ibx1,     iby1 + 1, ibx2,     iby2 + 1, outline, left, top, right, bottom);
        RenderUtils.fillTriangle(g, ipx,     ipy,     ibx1,     iby1,     ibx2,     iby2,     color,   left, top, right, bottom);
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/client/map/render/MapRenderers.kt
package net.shiroha233.roadweaver.client.map.render

import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.shiroha233.roadweaver.helpers.Records
import java.util.function.IntUnaryOperator

object MapRenderers {
    fun interface SegmentInView {
        fun test(x1: Int, z1: Int, x2: Int, z2: Int): Boolean
    }

    @JvmStatic
    fun renderGrid(
        g: GuiGraphics,
        font: Font,
        mapX: Int,
        mapY: Int,
        mapW: Int,
        mapH: Int,
        innerPad: Int,
        viewMinX: Double,
        viewMaxX: Double,
        viewMinZ: Double,
        viewMaxZ: Double,
        colorGrid: Int,
        gridTargetPx: Int,
        colorText: Int
    ) {
        GridRenderer.render(
            g,
            font,
            mapX,
            mapY,
            mapW,
            mapH,
            innerPad,
            viewMinX,
            viewMaxX,
            viewMinZ,
            viewMaxZ,
            colorGrid,
            gridTargetPx,
            colorText
        )
    }

    @JvmStatic
    fun renderStructures(
        g: GuiGraphics,
        points: List<BlockPos>,
        toScreenX: IntUnaryOperator,
        toScreenY: IntUnaryOperator,
        isInViewWorld: java.util.function.BiPredicate<Int, Int>,
        size: Int,
        color: Int,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int
    ) {
        for (p in points) {
            val bx = p.x
            val bz = p.z
            if (!isInViewWorld.test(bx, bz)) continue
            val x = toScreenX.applyAsInt(bx)
            val y = toScreenY.applyAsInt(bz)
            RenderUtils.drawPoint(g, x, y, size, color, left, top, right, bottom)
        }
    }

    @JvmStatic
    fun renderConnections(
        g: GuiGraphics,
        connections: List<Records.StructureConnection>,
        segmentInView: SegmentInView,
        toScreenX: IntUnaryOperator,
        toScreenY: IntUnaryOperator,
        thickness: Int,
        colorPlanned: Int,
        colorGenerating: Int,
        colorCompleted: Int,
        colorFailed: Int,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int
    ) {
        for (c in connections) {
            val fx = c.from.x
            val fz = c.from.z
            val tx = c.to.x
            val tz = c.to.z
            if (!segmentInView.test(fx, fz, tx, tz)) continue
            val x1 = toScreenX.applyAsInt(fx)
            val y1 = toScreenY.applyAsInt(fz)
            val x2 = toScreenX.applyAsInt(tx)
            val y2 = toScreenY.applyAsInt(tz)
            when (c.status) {
                Records.ConnectionStatus.PLANNED -> {
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, colorPlanned, thickness, left, top, right, bottom)
                }
                Records.ConnectionStatus.GENERATING -> {
                    RenderUtils.drawThickDashedLine(g, x1, y1, x2, y2, colorGenerating, thickness, 8, 6, left, top, right, bottom)
                }
                Records.ConnectionStatus.COMPLETED -> {
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, colorCompleted, thickness, left, top, right, bottom)
                }
                Records.ConnectionStatus.FAILED -> {
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, colorFailed, thickness, left, top, right, bottom)
                }
            }
        }
    }

    @JvmStatic
    fun renderRoadPolylines(
        g: GuiGraphics,
        polylines: List<List<BlockPos>>,
        segmentInView: SegmentInView,
        toScreenX: IntUnaryOperator,
        toScreenY: IntUnaryOperator,
        thickness: Int,
        color: Int,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int,
        lodStepBlocks: Int
    ) {
        for (pl in polylines) {
            if (pl.size < 2) continue
            var keep = pl[0]
            for (i in 1 until pl.size) {
                val curr = pl[i]
                val dx = curr.x - keep.x
                val dz = curr.z - keep.z
                val adx = kotlin.math.abs(dx)
                val adz = kotlin.math.abs(dz)
                val manhattan = adx + adz
                if (manhattan < lodStepBlocks) continue

                val x1w = keep.x
                val z1w = keep.z
                val x2w = curr.x
                val z2w = curr.z

                if (segmentInView.test(x1w, z1w, x2w, z2w)) {
                    val x1 = toScreenX.applyAsInt(x1w)
                    val y1 = toScreenY.applyAsInt(z1w)
                    val x2 = toScreenX.applyAsInt(x2w)
                    val y2 = toScreenY.applyAsInt(z2w)
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, color, thickness, left, top, right, bottom)
                }
                keep = curr
            }
            val tail = pl[pl.size - 1]
            if (!tail.equals(keep)) {
                val x1w = keep.x
                val z1w = keep.z
                val x2w = tail.x
                val z2w = tail.z
                if (segmentInView.test(x1w, z1w, x2w, z2w)) {
                    val x1 = toScreenX.applyAsInt(x1w)
                    val y1 = toScreenY.applyAsInt(z1w)
                    val x2 = toScreenX.applyAsInt(x2w)
                    val y2 = toScreenY.applyAsInt(z2w)
                    RenderUtils.drawThickLine(g, x1, y1, x2, y2, color, thickness, left, top, right, bottom)
                }
            }
        }
    }

    @JvmStatic
    fun renderLegend(
        g: GuiGraphics,
        font: Font,
        rightBound: Int,
        startY: Int,
        gap: Int,
        colorText: Int,
        colorStruct: Int,
        colorPlanned: Int,
        colorGenerating: Int,
        colorCompleted: Int,
        colorFailed: Int,
        structuresCount: Int,
        plannedCount: Int,
        generatingCount: Int,
        completedCount: Int,
        failedCount: Int
    ) {
        var y = startY

        val l1 = Component.translatable("gui.roadweaver.map.legend.structures").append(": ").append(structuresCount.toString())
        val w1 = font.width(l1)
        val x1 = rightBound - w1
        var sr = x1 - gap
        g.fill(sr - 5, y + 1, sr, y + 6, colorStruct)
        g.drawString(font, l1, x1, y, colorText, false)

        y += 16
        val l2 = Component.translatable("gui.roadweaver.map.legend.planned").append(": ").append(plannedCount.toString())
        val w2 = font.width(l2)
        val x2 = rightBound - w2
        sr = x2 - gap
        g.fill(sr - 28, y + 2, sr, y + 7, colorPlanned)
        g.drawString(font, l2, x2, y, colorText, false)

        y += 16
        val l3 = Component.translatable("gui.roadweaver.map.legend.generating").append(": ").append(generatingCount.toString())
        val w3 = font.width(l3)
        val x3 = rightBound - w3
        sr = x3 - gap
        val cy = y + 4
        RenderUtils.drawThickDashedLine(g, sr - 28, cy, sr, cy, colorGenerating, 5, 8, 6, sr - 28, y + 1, sr, y + 8)
        g.drawString(font, l3, x3, y, colorText, false)

        y += 16
        val l4 = Component.translatable("gui.roadweaver.map.legend.completed").append(": ").append(completedCount.toString())
        val w4 = font.width(l4)
        val x4 = rightBound - w4
        sr = x4 - gap
        g.fill(sr - 28, y + 2, sr, y + 7, colorCompleted)
        g.drawString(font, l4, x4, y, colorText, false)

        y += 16
        val l5 = Component.translatable("gui.roadweaver.map.legend.failed").append(": ").append(failedCount.toString())
        val w5 = font.width(l5)
        val x5 = rightBound - w5
        sr = x5 - gap
        g.fill(sr - 28, y + 2, sr, y + 7, colorFailed)
        g.drawString(font, l5, x5, y, colorText, false)
    }

    @JvmStatic
    fun drawPlayerArrow(
        g: GuiGraphics,
        sx: Int,
        sy: Int,
        yawDeg: Float,
        tipLen: Int,
        baseLen: Int,
        baseHalfWidth: Int,
        color: Int,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int,
        pxPerBlockX: Double,
        pxPerBlockZ: Double
    ) {
        val rx = Math.toRadians(yawDeg.toDouble())
        val vx = -kotlin.math.sin(rx)
        val vz = kotlin.math.cos(rx)

        var dirX = vx * pxPerBlockX
        var dirY = vz * pxPerBlockZ
        var mag = kotlin.math.hypot(dirX, dirY)
        if (mag < 1e-6) {
            dirX = 1.0
            dirY = 0.0
            mag = 1.0
        }
        dirX /= mag
        dirY /= mag

        val px = sx + dirX * tipLen
        val py = sy + dirY * tipLen
        val bx = sx - dirX * baseLen
        val by = sy - dirY * baseLen
        val perpX = -dirY
        val perpY = dirX

        val bx1 = bx + perpX * baseHalfWidth
        val by1 = by + perpY * baseHalfWidth
        val bx2 = bx - perpX * baseHalfWidth
        val by2 = by - perpY * baseHalfWidth

        val ipx = kotlin.math.round(px).toInt()
        val ipy = kotlin.math.round(py).toInt()
        val ibx1 = kotlin.math.round(bx1).toInt()
        val iby1 = kotlin.math.round(by1).toInt()
        val ibx2 = kotlin.math.round(bx2).toInt()
        val iby2 = kotlin.math.round(by2).toInt()

        val outline = 0xFFFFFFFF.toInt()
        RenderUtils.fillTriangle(g, ipx - 1, ipy, ibx1 - 1, iby1, ibx2 - 1, iby2, outline, left, top, right, bottom)
        RenderUtils.fillTriangle(g, ipx + 1, ipy, ibx1 + 1, iby1, ibx2 + 1, iby2, outline, left, top, right, bottom)
        RenderUtils.fillTriangle(g, ipx, ipy - 1, ibx1, iby1 - 1, ibx2, iby2 - 1, outline, left, top, right, bottom)
        RenderUtils.fillTriangle(g, ipx, ipy + 1, ibx1, iby1 + 1, ibx2, iby2 + 1, outline, left, top, right, bottom)
        RenderUtils.fillTriangle(g, ipx, ipy, ibx1, iby1, ibx2, iby2, color, left, top, right, bottom)
    }
}
